### PR TITLE
Using map to store the listener in ConnectionHandlerImpl

### PR DESCRIPTION
**Pull request by soulxu**
Commit Message: Using map to store the listener in ConnectionHandlerImpl
Additional Description:
Risk Level: low
Testing: unittest
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: no

Signed-off-by: He Jie Xu <hejie.xu@intel.com>

### CODE REVIEW

File: test/server/connection_handler_test.cc
Comment (soulxu) [inline test/server/connection_handler_test.cc:790]: A note for the reviewer, the original test ensure the first listener matched since the original code using the `list`. After using the unorder map, so this behavior changed. I assume the `MatchFirst` isn't part of API contract from the doc.

File: source/server/connection_handler_impl.cc
Comment (wbpcode) [inline source/server/connection_handler_impl.cc:92]: Can `absl::flat_hash_map` guarantee the stability of value address? May we should use `absl::node_hash_map` for `listener_map_by_tag_` or wrapper the details with `unique_ptr`? 🤔

File: source/server/connection_handler_impl.cc
Comment (wbpcode) [inline source/server/connection_handler_impl.cc:102]: IMO, it is not a good idea to use addresses to distinguish between old and new values. Because the memory may be reallocated. 
May be we should store the related listener tag the the `listener_map_by_address_`.

File: source/server/connection_handler_impl.cc
Comment (wbpcode) [inline source/server/connection_handler_impl.cc:143]: `listener_map_by_tag_` ?

File: source/server/connection_handler_impl.cc
Comment (wbpcode) [inline source/server/connection_handler_impl.cc:151]: `listener_map_by_tag_` ? the old `listeners_` list includes all the listeners and may be here we should iterate `listener_map_by_tag_` to replace original logic.

File: source/server/connection_handler_impl.cc
Comment (wbpcode) [inline source/server/connection_handler_impl.cc:158]: ditto.

File: source/server/connection_handler_impl.cc
Comment (wbpcode) [inline source/server/connection_handler_impl.cc:165]: ditto.

File: source/server/connection_handler_impl.cc
Comment (wbpcode) [inline source/server/connection_handler_impl.cc:172]: nit: asStringView()

**Comment by wbpcode** [comment]
Thanks. It's great and added some comments. cc @rojkov

File: source/server/connection_handler_impl.cc
Comment (soulxu) [reply source/server/connection_handler_impl.cc:92]: Oh, right, absl::flat_hash_map not guarantee that. thanks for point that out! I can use the unique_ptr, or std::reference_wrapper.

File: source/server/connection_handler_impl.cc
Comment (soulxu) [reply source/server/connection_handler_impl.cc:102]: got it.

File: source/server/connection_handler_impl.cc
Comment (soulxu) [reply source/server/connection_handler_impl.cc:143]: My final goal is to support multiple addresses listener. https://github.com/envoyproxy/envoy/issues/11184, 

The `listener_map_by_tag_`  will nest with a vector. For just iterate all the listeners, I just use the listener_map_by_address_ for more convenience, since it is just a flat map.

File: source/server/connection_handler_impl.cc
Comment (soulxu) [reply source/server/connection_handler_impl.cc:172]: got it

**Comment by soulxu** [comment]
@wbpcode thanks for the review!

File: source/server/connection_handler_impl.cc
Comment (wbpcode) [reply source/server/connection_handler_impl.cc:143]: If we finally complete it, will there are multiple address index for the same listener? Then we may call `shutdownListener()` method of same listener multiple times ?

In fact we can make the `listener_map_by_tag_` more simple just `absl::flat_hash_map<uint64_t, ActiveListenerDetailsPtr> listeners_`.

The reason why the address is saved in the previous `listeners_` list is to support searching by address. Now that we have a `listener_map_by_address_`, it is no longer necessary.

**Comment by wbpcode** [comment]
In addition, we may need distinguish the listener of different protocols.  Check https://github.com/envoyproxy/envoy/pull/17414.

File: source/server/connection_handler_impl.cc
Comment (soulxu) [reply source/server/connection_handler_impl.cc:143]: > If we finally complete it, will there are multiple address index for the same listener? Then we may call `shutdownListener()` method of same listener multiple times?

The different address will have different listener. so it won't be called multiple times with shutdownListener().

But I think I still should use `listener_map_by_tag_`, since `listener_map_by_address_` only store the newest listener, the drained listener will be overridden at `addListener`. And the ListenerManager will stop the listener first, then add new listener, but I shouldn't depend on how `ListenerManager` invoke `ConnectionhandlerImpl`, that isn't part of API contract.

> 
> In fact we can make the `listener_map_by_tag_` more simple just `absl::flat_hash_map<uint64_t, ActiveListenerDetailsPtr> listeners_`.

Yea, That is much simple, I plan to add tag and address instance field into `ActiveListenerDetail`

> 
> The reason why the address is saved in the previous `listeners_` list is to support searching by address. Now that we have a `listener_map_by_address_`, it is no longer necessary.

**Comment by soulxu** [comment]
> In addition, we may need distinguish the listener of different protocols. Check #17414.

thanks for another good point!

File: source/server/connection_handler_impl.cc
Comment (soulxu) [reply source/server/connection_handler_impl.cc]: @rojkov in the end, I use flat_hash_map + unique_ptr, let me know if you still think `node_hash_map` is a better option.

File: source/server/connection_handler_impl.cc
Comment (soulxu) [inline source/server/connection_handler_impl.cc:195]: I'm thinking to change this method name as `findInternalListenerByAddress`, since the assertion as below, this method is only used for internal listener.

File: source/server/connection_handler_impl.cc
Comment (soulxu) [inline source/server/connection_handler_impl.cc:244]: also think about change this method name to getBalancedTCPHandlerByAddress, since this method is only used for TCP listener

**Comment by jmarantz** [comment]
After @wbpcode approves I'll assign to a maintainer for the next pass. Thanks!

File: source/server/connection_handler_impl.cc
Comment (wbpcode) [inline source/server/connection_handler_impl.cc:48]: nit:
```
if(auto iter = listener_map_by_tag_.find(overridden_listener.value()); iter != listener_map_by_tag_.end()) {
  // ...
}
```

File: source/server/connection_handler_impl.cc
Comment (wbpcode) [inline source/server/connection_handler_impl.cc:60]: ditto.

File: source/server/connection_handler_impl.cc
Comment (wbpcode) [inline source/server/connection_handler_impl.cc:103]:
```suggestion
  listener_map_by_tag_.emplace(config.listenerTag(), details);
  // This map only store the new listener.
  if (absl::holds_alternative<std::reference_wrapper<ActiveTcpListener>>(
          details->typed_listener_)) {
    tcp_listener_map_by_address_.insert_or_assign(
        config.listenSocketFactory().localAddress()->asStringView(),
        std::move(details));
  } else if (absl::holds_alternative<std::reference_wrapper<ActiveInternalListener>>(
                 details->typed_listener_)) {
    internal_listener_map_by_address_.insert_or_assign(
        config.listenSocketFactory().localAddress()->asStringView(),
        std::move(details));
  }
```

File: source/server/connection_handler_impl.cc
Comment (wbpcode) [inline source/server/connection_handler_impl.cc:120]:
```suggestion
  auto listener_iter = listener_map_by_tag_.find(listener_tag);
  if (listener_iter != listener_map_by_tag_.end()) {
    absl::string_view address_view = listener_iter->second->address_->asStringView();
    // listener_map_by_address_ may already update to the new listener. Compare it with the one
    // which find from listener_map_by_tag_, only delete it when it is same listener.
    if (tcp_listener_map_by_address_.contains(address_view) &&
        tcp_listener_map_by_address_[address_view]
                ->listener_tag_ == listener_iter->second->listener_tag_) {
      tcp_listener_map_by_address_.erase(address_view);
    } else if (internal_listener_map_by_address_.contains(
                   address_view) &&
               internal_listener_map_by_address_[address_view]
                       ->listener_tag_ == listener_iter->second->listener_tag_) {
      internal_listener_map_by_address_.erase(address_view);
    }
```

File: source/server/connection_handler_impl.cc
Comment (wbpcode) [inline source/server/connection_handler_impl.cc:161]: when will `listener()` return nullptr? May be keep the original logic is ok?

File: source/server/connection_handler_impl.cc
Comment (wbpcode) [inline source/server/connection_handler_impl.cc:172]: ditto.

File: source/server/connection_handler_impl.cc
Comment (wbpcode) [inline source/server/connection_handler_impl.cc:181]: ditto.

File: source/server/connection_handler_impl.cc
Comment (wbpcode) [inline source/server/connection_handler_impl.cc:190]: ditto.

File: source/server/connection_handler_impl.cc
Comment (wbpcode) [inline source/server/connection_handler_impl.cc:265]: May be we can fold this to:

```
const std::string addr_str = address.ip()->version() == Network::Address::IpVersion::v4 ? ... : ... + std::to_string(address.ip()->port());
```

**Review by wbpcode — COMMENTED** [review]
This looks good for me overall. And just some minor comments.

File: source/server/connection_handler_impl.cc
Comment (wbpcode) [reply source/server/connection_handler_impl.cc:48]: Personal preference and feel free to just keep original code.

Comment (soulxu) [reply source/server/connection_handler_impl.cc:48]: done

File: source/server/connection_handler_impl.cc
Comment (soulxu) [reply source/server/connection_handler_impl.cc:103]: thanks, will update

File: source/server/connection_handler_impl.cc
Comment (soulxu) [reply source/server/connection_handler_impl.cc:120]: thanks, will update too!

File: source/server/connection_handler_impl.cc
Comment (soulxu) [reply source/server/connection_handler_impl.cc:161]: It can be nullptr when you try to stop the listener twice. Although the `shutdownListener` is just going to reset the unique pointer and that is fine, but I think it is better to have a check, seems we can't assume what `shutdownListener` will do in the future.

Comment (wbpcode) [reply source/server/connection_handler_impl.cc:161]: 🤔 SGTM.

Comment (soulxu) [reply source/server/connection_handler_impl.cc:161]: Seems the ActiveTCPListener have the check of listener is nullptr or not for `pauseListener` https://github.com/envoyproxy/envoy/blob/5f90ef2a615c07f7777342fc24496ff035fc6839/source/server/active_tcp_listener.cc#L113
but the ActiveUDPListener doesn't do that https://github.com/envoyproxy/envoy/blob/5f90ef2a615c07f7777342fc24496ff035fc6839/source/server/active_udp_listener.h#L97

File: source/server/connection_handler_impl.cc
Comment (soulxu) [reply source/server/connection_handler_impl.cc:265]: will do, thanks

**Comment by soulxu** [comment]
> This looks good for me overall. And just some minor comments.

thanks for the review!

**Comment by soulxu** [comment]
@wbpcode appreciate if you can take a look at again!

File: source/server/connection_handler_impl.cc
Comment (wbpcode) [inline source/server/connection_handler_impl.cc:171]: nit: looks like only these codes changed the original logic. can we just add some simple cases to cover it? (It doesn't look like this code affects security, so this is just a small suggestion. )

File: source/server/connection_handler_impl.cc
Comment (wbpcode) [inline source/server/connection_handler_impl.cc:224]: nit: `if (auto iter = listener_map_by_tag_.find(listener_tag); iter != listener_map_by_tag_.end()) {}`

File: source/server/connection_handler_impl.cc
Comment (wbpcode) [inline source/server/connection_handler_impl.cc:247]:
```
  if (auto listener_it = tcp_listener_map_by_address_.find(address->asStringView());
      listener_it != tcp_listener_map_by_address_.end()) {}
```

**Review by wbpcode — APPROVED** [review]
Thanks. LGTM. cc @jmarantz @envoyproxy/envoy-maintainers 

Some nit comments. But they are not worth blocking the review.

File: source/server/connection_handler_impl.cc
Comment (soulxu) [reply source/server/connection_handler_impl.cc:171]: done, thanks

File: source/server/connection_handler_impl.cc
Comment (ggreenway) [inline source/server/connection_handler_impl.cc:119]: It's more efficient to erase with `listener_iter` instead of by tag, so that we don't have to find the element again.

File: source/server/connection_handler_impl.cc
Comment (ggreenway) [inline source/server/connection_handler_impl.cc:152]: How can it be in this map, and have a nullptr listener?

File: source/server/connection_handler_impl.cc
Comment (ggreenway) [inline source/server/connection_handler_impl.cc:251]: Can this be `Envoy::OptRef`?

File: test/server/connection_handler_test.cc
Comment (ggreenway) [inline test/server/connection_handler_test.cc:531]: What would cause Envoy to stop or disable an already stopped listener? Please add a comment for the test explaining.

File: test/server/connection_handler_test.cc
Comment (ggreenway) [inline test/server/connection_handler_test.cc:776]: period at end of sentence

File: test/server/connection_handler_test.cc
Comment (ggreenway) [inline test/server/connection_handler_test.cc:793]: Capitalize `stop`, and period at end of sentence.

File: test/server/connection_handler_test.cc
Comment (ggreenway) [inline test/server/connection_handler_test.cc:813]: period at end

File: test/server/connection_handler_test.cc
Comment (ggreenway) [inline test/server/connection_handler_test.cc:766]: Please add comment on what config change operations would result in this sequence of listener operations.

**Review by ggreenway — CHANGES_REQUESTED** [review]
This looks really good. A few questions, and some comments to add. Thanks for fixing this scalability bottleneck!

/wait

File: source/server/connection_handler_impl.cc
Comment (soulxu) [reply source/server/connection_handler_impl.cc:119]: got it

File: source/server/connection_handler_impl.cc
Comment (soulxu) [reply source/server/connection_handler_impl.cc:152]: We will stop the listener before draining it,  the shutdownListener() https://github.com/envoyproxy/envoy/pull/19362/files#diff-fe538327af200510b78cadc759f2c3af989c822a370d3f8a005b92bc59568aa5R153 will reset the pointer. Then it is going to be nullptr.  That means if we shutdown twice, then the second time you will see a nullptr.

File: source/server/connection_handler_impl.cc
Comment (soulxu) [reply source/server/connection_handler_impl.cc:251]: yes, it can be

File: test/server/connection_handler_test.cc
Comment (soulxu) [reply test/server/connection_handler_test.cc:531]: Actually, we don't have such case in the current code(as I known), I just make it works in case we have such in the future. 
In tcp listener, there is check to keep the thing safe 
https://github.com/envoyproxy/envoy/blob/35d18c932a325bf62a67b2d95021a18e0e32d167/source/server/active_tcp_listener.cc#L112-L116

But udp listener hasn't such check
https://github.com/envoyproxy/envoy/blob/35d18c932a325bf62a67b2d95021a18e0e32d167/source/server/active_udp_listener.h#L97

So I add the check at ConnectionhandleImpl.

Or I should using the assertion and cleanup the tcp and udp listener check?

File: test/server/connection_handler_test.cc
Comment (soulxu) [reply test/server/connection_handler_test.cc:766]: got it

**Comment by soulxu** [comment]
@ggreenway thanks for the review!

I addressed your comments except for https://github.com/envoyproxy/envoy/pull/19362#discussion_r788936941, but let me know if you expect the assertion used there.

File: test/server/connection_handler_test.cc
Comment (ggreenway) [reply test/server/connection_handler_test.cc:532]: I understand now. I think this is fine as is.

**Review by ggreenway — APPROVED** [review]
This looks good. Thanks!

**Comment by howardjohn** [comment]
When upgrading our Istio test suites to this version everything started going a little wonky. I am not 100% sure what's happening but it looks like listeners are getting mixed up - calls are going to the wrong listeners as far as I can tell This is happening with bind_to_port=false. We were able to isolate the regression to this commit.

cc @lambdai

**Comment by mattklein123** [comment]
It wouldn't surprise me if with bind_to_port = false we now have some ordering variability or something else. Feel free to revert this PR until we sort that out. cc @soulxu

**Comment by soulxu** [comment]
I will do some tests on `bind_to_port = false` case. Then let you know when I figure out what happened.

**Comment by soulxu** [comment]
> When upgrading our Istio test suites to this version everything started going a little wonky. I am not 100% sure what's happening but it looks like listeners are getting mixed up - calls are going to the wrong listeners as far as I can tell This is happening with bind_to_port=false. We were able to isolate the regression to this commit.
> 
> cc @lambdai

@howardjohn do you have more failure info, or which testcase are failing. I checked the test, we have few tests on `bind_to_port = false` case 

https://github.com/envoyproxy/envoy/blob/a844bafc3130667f79fc1846a2e4b25831bbb5c8/test/server/connection_handler_test.cc#L959

https://github.com/envoyproxy/envoy/blob/a844bafc3130667f79fc1846a2e4b25831bbb5c8/test/server/connection_handler_test.cc#L771

I only thought the 0 port address can be a problem. But that shouldn't the case, since I can't create multiple 0 port addresses with bind_to_port as false.

**Comment by howardjohn** [comment]
Not Envoy tests, but Istio tests. It's pretty much all of our tests (
https://github.com/istio/istio/pull/37004) since it seems all requests go
to random listeners

**Comment by soulxu** [comment]
I could be wrong here

https://github.com/envoyproxy/envoy/blob/a844bafc3130667f79fc1846a2e4b25831bbb5c8/source/server/connection_handler_impl.cc#L245-L249

I should check `listener_it->second->listener() != nullptr`, otherwise a draining listener could get new connection balanced to.

**Pull request merged**
